### ISSUE TITLE
`map_dict` with str -> nonstr mapping fails to map when `default=pl.first()`

### ISSUE THREAD

**Issue by mcrumiller** [pola-rs/polars#9841]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl

pl.Series(['a', 'b', 'c']).map_dict({'a': 1}, default=pl.first())
```
```
Series: '' [str]
[
        null  <--note that this is the only value that should successfully map
        "b"
        "c"
]
```

### Issue description

When using a `map_dict` from str to non-str, using `pl.first()` causes all successfully-mapped values to become null.

### Expected behavior

`a` should map to 1, the rest should map to `b` and `c`. We should either coerce the successful mapping to the string `"1"`, or should raise.

### Installed versions

<details>

```
--------Version info---------
Polars:              0.18.6
Index type:          UInt32
Platform:            Windows-10-10.0.19045-SP0
Python:              3.11.2 (tags/v3.11.2:878ead1, Feb  7 2023, 16:38:35) [MSC v.1934 64 bit (AMD64)]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
connectorx:          0.3.2a5
deltalake:           <not installed>
fsspec:              <not installed>
matplotlib:          3.7.1
numpy:               1.25.1
pandas:              2.0.3
pyarrow:             11.0.0
pydantic:            <not installed>
sqlalchemy:          2.0.7
xlsx2csv:            0.8.1
xlsxwriter:          3.0.9
```

</details>

**Comment by ProgramPundit** [comment]
Hi, can I work on this issue?. Can you please assign me to this issue?

**Comment by mcrumiller** [comment]
@Rohith-Manjunatha to add a little help here to get you started. The reason is in this section right here: https://github.com/pola-rs/polars/blob/main/py-polars/polars/expr/expr.py#L8744-L8748

```python
return_dtype_ = (
    df.lazy().select(default).dtypes[0]
    if return_dtype is None and isinstance(default, Expr)
    else return_dtype
)
```

If the `default` value is an expression (which it is in the case of `pl.first()`) then the return dtype is set to the series dtype. Thus here we have return dtype = `pl.Utf8`. The next step remaps the keys to `pl.Utf8`, which causes our valid map to be null and everything else to be mapped to a `str`:

```python
pl.Series([1], dtype=pl.Utf8)
```
```
shape: (1,)
Series: '' [str]
[
	null
]
```

The best course of action here might be instead to create the keys series without specifying dtype and then attempting to `.cast()` instead. It looks like this code was added in #7797 by @ghuls--Gert, care to chime in here? A few thoughts:

* the dtype of the mapping and the default expression do not match. Should this simply raise immediately? The issue here is that int.cast(str) works fine, but `pl.Series(X, dtype=str)` fails if X is ints.
* I encountered this because I had default set to `pl.first()`, but every element mapped successfully, meaning `pl.first()` was never called, and so I fully expected my output dtype NOT to be `Utf8` but instead to be an int, as specified by the mapping. Instead, every single element ended up null.
